016-kafka-connect-build.md

```diff
@@ -0,0 +1,137 @@
+
+The second part - `output` - will configure how the newly built image will be handled.
+It will support two types:
+* `docker` to push the image to any Docker compatible container repository.
+* `imagestream` to push the image to OpenShift ImageStream (supported only on OpenShift)
+
+Strimzi will not run its own container repository.
+So users of the `docker` type build will need to provide credentials to their own Docker compatible container registry.
+This does not have to run within the same Kubernetes cluster - it can be also SaaS registry such as Docker Hub or Quay.io.
+
+Following is an example of the Kafka Connect CR:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+metadata:
+  name: my-connect-cluster
+  annotations:
+    strimzi.io/use-connector-resources: "true"
+spec:
+  version: 2.6.0
+  replicas: 1
+  bootstrapServers: my-cluster-kafka-bootstrap:9093
+  build:
+    output:
+      type: docker
+      image: docker.io/user/image:tag
+      pushSecret: dockerhub-credentaials
+    plugins:
+      - name: my-connector
+        artifacts:
+          - type: jar
+            url: https://some.url/artifact.jar
+            sha512sum: abc123
+          - type: jar
+            url: https://some.url/artifact2.jar
+            sha512sum: def456
+      - name: my-connector2
+        # ...
+  tls:
+    trustedCertificates:
+      - secretName: my-cluster-cluster-ca-cert
+        certificate: ca.crt
+  config:
+    group.id: connect-cluster
+    offset.storage.topic: connect-cluster-offsets
+    config.storage.topic: connect-cluster-configs
+    status.storage.topic: connect-cluster-status
+```
+
+When the `build` section is specified in the `KafkaConnect` resource, Strimzi will generate `Dockerfile` which will correspond to the configured connectors and their artifacts.
+It will build the container image from this `Dockerfile` in a separate pod and push it to the configured output (Docker registry / ImageStream).
+It will take the digest of this image and configure the Kafka Connect cluster to use this image.
+
+The build will use two different technologies:
+* [Kaniko](https://github.com/GoogleContainerTools/kaniko) will be used on Kubernetes
+* OpenShift Builds will be used on OKD / OpenShift 
+
+Kaniko is a daemon-less container builder which does not require any special privileges and runs in user-space.
+Kaniko is just a container to which we pass the generated `Dockerfile` and let it build the container and push it to the registry.
+It does not require any special installation - neither from the user nor from Strimzi it self. 
+We just use the container.
+
+OpenShift Builds are part of the OKD / OpenShift Kubernetes platform and also do not require any special privileges.
+So it should work for most users across wide variate of environments.
+
+## Affected/not affected projects
+
+The existing `KafkaConnectS2I` resource is not affected by it.
+But it should be deprecated and later removed.
```

Comment:
Using should here is ambiguous. Unless this proposal is deprecating it then it's best not to mention it. If it is proposing deprecation then it should say so unambiguously: "KafkaConnectS2I will be deprecated."

Reply:
I see this proposal as superior, so I think we should deprecate the `KafkaconnectS2I` (and remove later of course) it once it is implemented. I updated the proposal accordingly.